### PR TITLE
Add function to query with initial timestamp.

### PR DESCRIPTION
**Pull request by daigorowhite**
In case of large table and not needed all record, or user just need recent record and realtime data from now, initial query is required with specific timestamp. but currently, initial query uses `Timestamp(0L)`.

In this PR , `kafka-connect-jdbc`'s `source` have new config which is `timestamp.initial`.
if you specify `timestamp.initial`, it works like this. 

- `0` : `1970-01-01 00:00:00.000` (Same as existed)
- `-1` : Current time will be used 
- `1528858322371` : `2018-06-13 11:52:02.371` ( UnixTime milliseconds )

and default is 0.

In our case, we need this function, could you have a look, please?

### CODE REVIEW

**Comment by daigorowhite** [comment]
[clabot:check]

**Comment by daigorowhite** [comment]
hi @rhauch 
I think this use case is needed for some users,
since this issue has same topic for `incrementing` mode. ( our PR is currently only for `timestamp`)
https://github.com/confluentinc/kafka-connect-jdbc/issues/432

If you have time , please have a look this PR.

thanks

File: src/test/java/io/confluent/connect/jdbc/source/JdbcSourceTaskUpdateTest.java [hunk 1]
```diff
@@ -472,6 +551,14 @@ private void startTask(String timestampColumn, String incrementingColumn, String
...
 
   private void startTask(String timestampColumn, String incrementingColumn, String query, Long delay) {
+    startTask(timestampColumn, incrementingColumn, query, delay, null, null);
+  }
+
+  private void startTask(String timestampColumn, String incrementingColumn, String query, Long delay, Long timestampInitial) {
+    startTask(timestampColumn, incrementingColumn, query, delay, timestampInitial, null);
+  }
+
+  private void startTask(String timestampColumn, String incrementingColumn, String query, Long delay, Long timestampInitial, Long incrementInitial) {
```
Comment (rhauch) [inline src/test/java/io/confluent/connect/jdbc/source/JdbcSourceTaskUpdateTest.java]: Is `incrementInitial` an artifact of trying to do something similar for the incremental offset? It's currently unused, so we should remove it or add support for incremental offsets, too.

File: docs/source_config_options.rst [hunk 2]
```diff
@@ -130,6 +130,13 @@ Mode
   * Default: ""
   * Importance: medium
 
+``timestamp.initial``
+  Unix timestamp milliseconds. If specified, initial query will use this parameter as a timestamp start offset. If you set -1, current timestamp will be used.
+
+  * Type: long
+  * Default: 0
+  * Importance: low
+
```
Comment (rhauch) [inline docs/source_config_options.rst]: This file was removed from the codebase.

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java [hunk 3]
```diff
@@ -165,6 +165,14 @@
   public static final String TIMESTAMP_COLUMN_NAME_DEFAULT = "";
   private static final String TIMESTAMP_COLUMN_NAME_DISPLAY = "Timestamp Column Name";
 
+  public static final String TIMESTAMP_INITIAL_CONFIG = "timestamp.initial";
+  public static final long TIMESTAMP_INITIAL_DEFAULT = 0;
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java]: Should the default be `null`? That seems to be better than `0` if we want this to be an optional property.

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java [hunk 4]
```diff
@@ -165,6 +165,14 @@
...
+  public static final long TIMESTAMP_INITIAL_CURRENT = -1;
+  public static final String TIMESTAMP_INITIAL_DOC =
+      "Long value to specify when the query offset starts."
+      + "offset starts from current time if the value is -1.";
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java]: How about:
```
"The epoch timestamp used for initial queries that use timestamp criteria. "
      + "Use -1 to use the current time. If not specified, all data will be retrieved."
```

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java [hunk 5]
```diff
@@ -163,7 +166,10 @@ public void start(Map<String, String> properties) {
         default:
           throw new ConnectException("Unexpected query mode: " + queryMode);
       }
+
       Map<String, Object> offset = offsets == null ? null : offsets.get(partition);
+      offset = offset == null ? fetchInitialOffset(timestampInitial) : offset;
+
```
[the remark this answers is not in the document]
Comment (rhauch) [reply src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: Actually, we should use `dialect.currentTimeOnDB(connection, calendar)` to get the current time in the database. Also, be sure to see the recent changes from #505 that allow the user to specify the time zone, and that will affect the calendar that's used.

**Comment by daigorowhite** [comment]
@rhauch thanks for having review!
I have handled your comment and modify the code.
please check again ? and sorry, I have no idea about jenkins checkstyle...
could you tell me how to fix it ?

**Comment by daigorowhite** [comment]
Let me try to remind again @rhauch 
Thanks

**Comment by rgogulakrishnan** [comment]
Hi, any update on this feature..? When this will be rolled out..

**Comment by daigorowhite** [comment]
hi , I would like to merge this since I got approval already. no update now, I am just waiting :)

**Comment by remis87** [comment]
Any updates on this feature yet? Waiting for it passionately

**Comment by aliasbadwolf** [comment]
@daigorowhite Although I like the idea to handle it through configuration (and I am super late to comment) but the same can be done by setting the initial timestamp in the offset topic for connect framework. Before starting the connector, just add an entry to the offset topic with whatever timestamp you want to run the initial query from (and this can handle both timestamp and incremental scenarios). Although its a hack, until your changes get merged.

**Comment by HunderlineK** [comment]
@rhauch @wicknicks Could you please review the PR again and merge it if it looks okay? Thanks

**Comment by sohaibfarooqi** [comment]
Any updates on this? Waiting for this feature for quite some time!

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java [hunk 6]
```diff
@@ -256,6 +260,36 @@ public void start(Map<String, String> properties) {
...
+      return partitionOffset;
+    } else {
+      Map<String, Object> initialPartitionOffset = null;
+      // no offsets found
+      Long timestampInitial = config.getLong(JdbcSourceConnectorConfig.TIMESTAMP_INITIAL_CONFIG);
+      if (timestampInitial != null) {
+        // start at the specified timestamp
+        if (timestampInitial == JdbcSourceConnectorConfig.TIMESTAMP_INITIAL_CURRENT) {
+          // use the current time
+          try (Connection con = cachedConnectionProvider.getConnection()) {
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: Can you pull the `getConnection` outside of the try-with-resources?
It's a cached connection object, and closing it here invalidates that object elsewhere and forces the task to reconnect before continuing. See how `validateNonNullable` does it for an example.

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java [hunk 7]
```diff
@@ -38,16 +46,9 @@
...
 import java.util.concurrent.atomic.AtomicBoolean;
 
-import io.confluent.connect.jdbc.dialect.DatabaseDialect;
-import io.confluent.connect.jdbc.dialect.DatabaseDialects;
-import io.confluent.connect.jdbc.util.CachedConnectionProvider;
-import io.confluent.connect.jdbc.util.ColumnDefinition;
-import io.confluent.connect.jdbc.util.ColumnId;
-import io.confluent.connect.jdbc.util.TableId;
-import io.confluent.connect.jdbc.util.Version;
-
```
[the remark this answers is not in the document]
Comment (daigorowhite) [reply src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: I reverted.

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java [hunk 8]
```diff
@@ -149,6 +150,8 @@ public void start(Map<String, String> properties) {
         = config.getLong(JdbcSourceTaskConfig.TIMESTAMP_DELAY_INTERVAL_MS_CONFIG);
     boolean validateNonNulls
         = config.getBoolean(JdbcSourceTaskConfig.VALIDATE_NON_NULL_CONFIG);
+    Long timestampInitial
+        = config.getLong(JdbcSourceConnectorConfig.TIMESTAMP_INITIAL_CONFIG);
```
[the remark this answers is not in the document]
Comment (daigorowhite) [reply src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: I removed this one .

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java (hunk 6 above)
Comment (daigorowhite) [reply src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: got it, I moved this one. Could you check my change if I understand your comment ?

**Comment by daigorowhite** [comment]
@gharris1727 
Hi, Thank you for your kind review! I have fixed with your comment.
and, I switched destination `5.5.x` .
please kindly review again?

**Pull request merged**